src/lib/services/access-service.ts

```diff
@@ -662,32 +679,63 @@ export class AccessService {
         }
 
         await this.validateRole(role, role.id);
+        const existingRole = await this.roleStore.get(role.id);
         const baseRole = {
             id: role.id,
             name: role.name,
             description: role.description,
             roleType,
         };
         const rolePermissions = role.permissions;
-        const newRole = await this.roleStore.update(baseRole);
+        const updatedRole = await this.roleStore.update(baseRole);
+        const existingPermissions = await this.store.getPermissionsForRole(
+            role.id,
+        );
         if (rolePermissions) {
-            await this.store.wipePermissionsFromRole(newRole.id);
+            await this.store.wipePermissionsFromRole(updatedRole.id);
             if (roleType === CUSTOM_ROOT_ROLE_TYPE) {
                 await this.store.addPermissionsToRole(
-                    newRole.id,
+                    updatedRole.id,
                     rolePermissions,
                 );
             } else {
                 await this.store.addEnvironmentPermissionsToRole(
-                    newRole.id,
+                    updatedRole.id,
                     rolePermissions,
                 );
             }
         }
-        return newRole;
+        const updatedPermissions = await this.store.getPermissionsForRole(
+            role.id,
+        );
+        this.eventService.storeEvent({
+            type: ROLE_UPDATED,
+            createdBy: role.createdBy || 'unknown',
+            data: {
+                ...updatedRole,
+                permissions: this.sanitizePermissions(updatedPermissions),
+            },
+            preData: {
+                ...existingRole,
+                permissions: this.sanitizePermissions(existingPermissions),
+            },
+        });
+        return updatedRole;
+    }
+
+    sanitizePermissions(
+        permissions: IPermission[],
+    ): { name: string; environment?: string }[] {
+        return permissions.map(({ name, environment }) => {
+            const sanitizedEnvironment =
+                environment && environment !== null && environment !== ''
+                    ? environment
+                    : undefined;
+            return { name, environment: sanitizedEnvironment };
+        });
     }
 
-    async deleteRole(id: number): Promise<void> {
+    async deleteRole(id: number, deletedBy = 'unknown'): Promise<void> {
```

Comment:
in the enterprise repo we should me this transactional so that we guarantee event storage

Reply:
I'll do that as part of passing down the author of the action